runtime/src/secondary_index.rs

```diff
@@ -1,157 +1,95 @@
-use crate::contains::Contains;
 use dashmap::{mapref::entry::Entry::Occupied, DashMap};
-use log::*;
-use solana_sdk::{clock::Slot, pubkey::Pubkey};
-use std::{
-    borrow::Borrow,
-    collections::{hash_map, HashMap, HashSet},
-    fmt::Debug,
-    sync::{Arc, RwLock},
-};
+use solana_sdk::pubkey::Pubkey;
+use std::{collections::HashSet, fmt::Debug, sync::RwLock};
 
-pub type SecondaryReverseIndexEntry = RwLock<HashMap<Slot, Pubkey>>;
+// The only cases where an inner key should map to a different outer key is
+// if the key had different account data for the indexed key across different
+// slots. As this is rare, it should be ok to use a Vec here over a HashSet, even
+// though we are running some key existence checks.
+pub type SecondaryReverseIndexEntry = RwLock<Vec<Pubkey>>;
 
 pub trait SecondaryIndexEntry: Debug {
-    fn get_or_create(&self, key: &Pubkey, f: &dyn Fn(&RwLock<HashSet<Slot>>));
-    fn get<T>(&self, key: &Pubkey, f: &dyn Fn(Option<&RwLock<HashSet<Slot>>>) -> T) -> T;
-    fn remove_key_if_empty(&self, key: &Pubkey);
+    fn insert_if_not_exists(&self, key: &Pubkey);
+    // Removes a value from the set. Returns whether the value was present in the set.
+    fn remove_inner_key(&self, key: &Pubkey) -> bool;
     fn is_empty(&self) -> bool;
     fn keys(&self) -> Vec<Pubkey>;
     fn len(&self) -> usize;
 }
 
 #[derive(Debug, Default)]
 pub struct DashMapSecondaryIndexEntry {
-    pubkey_to_slot_set: DashMap<Pubkey, RwLock<HashSet<Slot>>>,
+    account_keys: DashMap<Pubkey, ()>,
 }
 
 impl SecondaryIndexEntry for DashMapSecondaryIndexEntry {
-    fn get_or_create(&self, key: &Pubkey, f: &dyn Fn(&RwLock<HashSet<Slot>>)) {
-        let slot_set = self.pubkey_to_slot_set.get(key).unwrap_or_else(|| {
-            self.pubkey_to_slot_set
-                .entry(*key)
-                .or_insert(RwLock::new(HashSet::new()))
-                .downgrade()
-        });
-
-        f(&slot_set)
-    }
-
-    fn get<T>(&self, key: &Pubkey, f: &dyn Fn(Option<&RwLock<HashSet<Slot>>>) -> T) -> T {
-        let slot_set = self.pubkey_to_slot_set.get(key);
-
-        f(slot_set.as_ref().map(|entry_ref| entry_ref.value()))
+    fn insert_if_not_exists(&self, key: &Pubkey) {
+        self.account_keys
+            .get(key)
+            .unwrap_or_else(|| self.account_keys.entry(*key).or_default().downgrade());
```

Comment:
Maybe a stupid question, why do we need this vs. just `self.account_keys.entry(*key).or_default()`

Reply:
Heh it's only necessary because I didn't clean this up as much as I could have 😃.

`get` returns a `Ref` and `.entry()` returns a `RefMut`, and the `unwrap_or_else` enforces the results must be of the same type.

Changed to check for `is_none()` instead here: https://github.com/solana-labs/solana/pull/17048/commits/1c43c59361923354816b28af61ef81d1c35a2911